contrib/verifybinaries/verify.py

```diff
@@ -59,124 +194,457 @@ def download_with_wget(remote_file, local_file=None):
     return result.returncode == 0, result.stdout.decode().rstrip()
 
 
+def download_lines_with_urllib(url) -> (bool, [str]):
+    """Get (success, text lines of a file) over HTTP."""
+    try:
+        return True, [
+            line.strip().decode() for line in urllib.request.urlopen(url).readlines()]
+    except urllib.request.HTTPError:
+        log.warning(f"HTTP request to {url} failed (HTTPError)")
+    except Exception as e:
+        log.warning(f"HTTP request to {url} failed ({e})")
+    return (False, [])
+
+
+def verify_with_gpg(
+    signature_filename,
+    output_filename: t.Optional[str] = None
+) -> t.Tuple[int, str]:
+    # Enforce English output since we're relying on output that contains natural language.
+    # Use string-shell form (vs. args as list) because passing LANGUAGE via `env` doesn't
+    # seem to work.
+    args = (
+        'LANGUAGE=en gpg --yes --decrypt '
+        + (f' --output {output_filename} ' if output_filename else '')
+        + signature_filename)
+
+    result = subprocess.run(args, shell=True, stderr=subprocess.STDOUT, stdout=subprocess.PIPE)
+    return result.returncode, result.stdout.decode().rstrip()
+
+
+def remove_files(filenames):
+    for filename in filenames:
+        os.remove(filename)
+
+
+class SigData:
+    """GPG signature data as parsed from GPG stdout."""
+    def __init__(self, key: str, name: str, trusted: bool, status: str):
+        self.key = key
+        self.name = name
+        self.trusted = trusted
+        self.status = status
+
+    def __repr__(self):
+        return (
+            "SigData(%r, %r, trusted=%s, status=%r)" %
+            (self.key, self.name, self.trusted, self.status))
+
+
+def parse_gpg_result(output: str) -> ([SigData], [SigData], [SigData]):
+    """Returns good, unknown, and bad signatures from GPG stdout."""
+    good_sigs = []
+    unknown_sigs = []
+    bad_sigs = []
+    total_resolved_sigs = 0
+    curr_key = None
+
+    for i, line in enumerate(output):
+        if 'using RSA key' in line:
```

Comment:
Would seem more readable and also more reassuring to do a fuller parse of GPG lines, instead of picking out little pieces of text. E.g.

```python
m = re.match(^" *using RSA key (0x[0-9a-f]{18}|[0-9a-f]{40}$")
```

If you're just doing fragments like `'using RSA key' in line`, who is to say the script won't match the text `NOT using RSA key` and you just gave me a fake binary and ran away with my secret keys.

Reply:
Excellent point, thanks!